### PR TITLE
add support for freeing C/C++ objects used purely within wrappers

### PR DESCRIPTION
**Pull request by cosnicolaou**
This PR addresses a problem we recently ran into when generating C bindings to allow go to be called from python. The C bindings generated for the go shared libraries will allocate strings on the C heap
in order for them to be returned to the python C binding. The python wrapper will typically make a copy of these via Py_BindValue but there is no mechanism for this memory to be freed since it never leaves the python wrapper. This PR adds a keyword `free_on_copy` which can be used to instruct the python wrapper to free this storage.

### CODE REVIEW

**Comment by gjcarneiro** [comment]
This is missing a test case.  I guess I can add it, or can try to update WAF to the latest version, whichever is faster, to be determined later.

**Comment by cosnicolaou** [comment]
FYI - I'm in the process of updating to the latest version of waf. Would you prefer that I land that in a separate PR, merge that and then add a test case to this PR. I think that's cleaner than merging it into this one, but I can do it either way,

**Comment by gjcarneiro** [comment]
Possibly a separate PR is better; thanks!

File: pybindgen/typehandlers/base.py [hunk 1]
```diff
@@ -1155,6 +1155,7 @@ def new(cls, *args, **kwargs):
                 args[0] = type_traits
                 args = tuple(args)
                 try:
+                    print(type_handler_class)
```
[the remark this answers is not in the document]
Comment (cosnicolaou) [reply pybindgen/typehandlers/base.py]: > examining the generated source code in the test is unorthodox, but I can accept it.
do you have any suggestions for how else to do this? In other circumstances I've written tests to check for memory leaks but those can be difficult to de-flake. I can give that a go too.

I will of course add the normal tests of the generated code.

[the remark this answers is not in the document]
Comment (cosnicolaou) [reply pybindgen/typehandlers/base.py]: PTAL now that I've added new tests etc. Thanks!

**Comment by gjcarneiro** [comment]
test_free_after_copy_leak seems potentially brittle, or dependant on implementation details on how system manages memory.  Leaving it in for now, but if it starts giving me CI failures, I'm afraid I'll probably just skip this test.

Otherwise, thank you!

**Comment by cosnicolaou** [comment]
I ran them a few dozen times locally and in a docker container but it seems that they fail immediately on travis, (I didn't notice the badge at the bottom of your readme originally), so if I get a bit more time I'll see if it's even possible to de-flake these. Sorry for that hassle!

**Comment by gjcarneiro** [comment]
I had to disable the flaky test, sorry.  It's fine, I'm happy with just the basic test.

**Pull request merged**